src/refdb_fs.c

```diff
@@ -466,10 +452,176 @@ static int ref_error_notfound(const char *name)
 	return GIT_ENOTFOUND;
 }
 
-static int packed_lookup(
-	git_reference **out,
-	refdb_fs_backend *backend,
-	const char *ref_name)
+static const char *packed_set_peeling_mode(
+        const char *data,
+        size_t data_sz,
+        refdb_fs_backend *backend)
+{
+	static const char *traits_header = "# pack-refs with:";
+	const char *eol;
+	backend->peeling_mode = PEELING_NONE;
+
+	if (data_sz == 0 || *data != '#') {
+		return data;
+	}
+
+	if (git__prefixncmp(data, data_sz, traits_header) == 0) {
+		size_t hdr_sz = strlen(traits_header);
+		const char *sorted = " sorted ";
+		const char *peeled = " peeled ";
+		const char *fully_peeled = " fully-peeled ";
+		data += hdr_sz;
+		data_sz -= hdr_sz;
+
+		eol = memchr(data, '\n', data_sz);
+
+		if (!eol)
+			return NULL;
+
+		if (git__memmem(data, eol - data, fully_peeled, strlen(fully_peeled))) {
+			backend->peeling_mode = PEELING_FULL;
+		} else if (git__memmem(data, eol - data, peeled, strlen(peeled))) {
+			backend->peeling_mode = PEELING_STANDARD;
+		}
+
+		if (git__memmem(data, eol - data, sorted, strlen(sorted))) {
+			backend->sorted = true;
+		} else {
+			backend->sorted = false;
+		}
+
+		return eol + 1;
+	}
+	return data;
+}
+
+static int packed_map_check(refdb_fs_backend *backend)
+{
+	int error = 0;
+	git_file fd = -1;
+	struct stat st;
+
+	if ((error = git_mutex_lock(&backend->prlock)) < 0) {
+		return error;
+	}
+
+	if (backend->packed_refs_map.data) {
+		git_mutex_unlock(&backend->prlock);
+		return error;
+	}
+
+	fd = git_futils_open_ro(backend->refcache->path);
+	if (fd < 0) {
+		git_mutex_unlock(&backend->prlock);
```

Comment:
I might have a goto here...  this is just about the size of a function where it has enough return points that reading through to make sure that we've unlocked the mutex on all of them is not _hard_, exactly, but a little meh.  Having a `goto done` or the like helps (me, at least) reason a little bit better about that.  Though this is rather stylistic, I concede.